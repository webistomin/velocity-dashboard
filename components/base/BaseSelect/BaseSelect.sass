+b('base-select')
  +e('label')
    display: block
    margin-bottom: rem(6px)

  +e('select')
    .vs__dropdown-toggle
      border: rem(1px) solid var(--color-lavender)
      border-radius: 5px
      padding: rem(3px) 0 rem(8px)

    .vs__selected-options
      padding-left: rem(16px)
      padding-right: rem(16px)

    .vs__dropdown-option
      padding-left: rem(24px)
      padding-right: rem(24px)

    .vs__dropdown-option--highlight
      background-color: var(--color-primary)
      color: var(--color-white)

    .vs__dropdown-menu
      padding: 0
      box-shadow: none
      border-radius: 5px
      border: rem(1px) solid var(--color-lavender)
      top: calc(100% + 5px)

    .vs__selected
      color: var(--color-primary)

    .vs__search::placeholder
      color: var(--color-gray-3)
