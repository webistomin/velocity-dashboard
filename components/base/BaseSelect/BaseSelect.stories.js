import { array, text } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';

import BaseSelect from './BaseSelect';

export default {
  title: 'BaseSelect',
  parameters: {
    options: {
      showPanel: true,
      selectedPanel: 'storybook/knobs/panel',
      previewTabs: {
        actions: {
          hidden: true,
        },
      },
    },
  },
};

export const Default = () => ({
  components: { BaseSelect },
  data() {
    return {
      selected: this.value,
    };
  },
  props: {
    options: {
      type: Array,
      default: array('options', ['1', '2', '3']),
      required: true,
    },
    value: {
      type: String,
      default: text('value', ''),
      required: true,
    },
    label: {
      type: String,
      default: text('label', 'Text'),
      required: true,
    },
    id: {
      type: String,
      default: text('id', 'base-form-group'),
      required: true,
    },
    placeholder: {
      type: String,
      default: text('placeholder', 'placeholder'),
    },
  },
  methods: {
    onInput: action('emit input'),
  },
  template: `<base-select
              :options='this.options'
              :label='this.label'
              :id='this.id'
              :placeholder='this.placeholder'
              @input='selected = $event, this.onInput'
              :value='this.selected'
  />`,
});
