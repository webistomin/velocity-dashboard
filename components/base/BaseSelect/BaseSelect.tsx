import { VueComponent } from 'types/vue-components';
import { Component, Prop, Emit } from 'nuxt-property-decorator';
import { VNode } from 'vue';

import VSelect from 'vue-select';

import './BaseSelect.sass';

export interface IBaseSelectProps {
  options: string[];
  value: string;
  onInput?: (value: string) => string;
  id: string;
  label: string;
  placeholder?: string;
}

@Component({
  name: 'BaseSelect',
  components: { VSelect },
})
export default class BaseSelect extends VueComponent<IBaseSelectProps> {
  @Prop({ required: true })
  private readonly options!: IBaseSelectProps['options'];

  @Prop({ required: true })
  private readonly value!: IBaseSelectProps['value'];

  @Prop({ required: true })
  private readonly id!: IBaseSelectProps['id'];

  @Prop({ required: true })
  private readonly label!: IBaseSelectProps['label'];

  @Prop()
  private readonly placeholder!: IBaseSelectProps['placeholder'];

  @Emit('input')
  public onInput(value: string): string {
    return value;
  }

  public render(): VNode {
    const { id, label, value, onInput, options, placeholder } = this;

    return (
      <div class='base-select'>
        <label for={id} class='base-select__label'>
          {label}
        </label>
        <VSelect
          value={value}
          onInput={onInput}
          options={options.sort()}
          inputId={id}
          clearable={false}
          placeholder={placeholder}
          class='base-select__select'>
          <template slot='open-indicator'>
            <span />
          </template>
        </VSelect>
      </div>
    );
  }
}
