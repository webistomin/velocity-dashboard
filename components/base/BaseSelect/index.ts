import BaseSelect from './BaseSelect';

export default BaseSelect;
