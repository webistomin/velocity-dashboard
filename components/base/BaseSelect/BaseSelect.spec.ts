import { mount } from '@vue/test-utils';
import BaseSelect from './BaseSelect';

describe('BaseSelect', () => {
  const selectSelector = '.base-select__select';

  const factory = (propsData = {}, options = {}) => {
    return mount(BaseSelect, {
      propsData: {
        options: ['1', '2', '3'],
        value: '',
        id: 'id',
        label: 'label',
        ...propsData,
      },
      stubs: ['VSelect'],
      ...options,
    });
  };

  it('Matches snapshot', () => {
    const wrapper = factory().html();
    expect(wrapper).toMatchSnapshot();
  });

  it('Emit "input" event', () => {
    const wrapper = factory();
    const input = wrapper.find(selectSelector);
    input.trigger('input');
    expect(wrapper.emitted().input);
  });
});
