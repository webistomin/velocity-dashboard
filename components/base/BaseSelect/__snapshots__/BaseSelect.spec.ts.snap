// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseSelect Matches snapshot 1`] = `
"<div class=\\"base-select\\"><label for=\\"id\\" class=\\"base-select__label\\">label</label>
  <div dir=\\"auto\\" class=\\"v-select vs--single vs--searchable base-select__select\\">
    <div id=\\"vs1__combobox\\" role=\\"combobox\\" aria-expanded=\\"false\\" aria-owns=\\"vs1__listbox\\" aria-label=\\"Search for option\\" class=\\"vs__dropdown-toggle\\">
      <div class=\\"vs__selected-options\\"> <input id=\\"id\\" aria-autocomplete=\\"list\\" aria-labelledby=\\"vs1__combobox\\" aria-controls=\\"vs1__listbox\\" type=\\"search\\" autocomplete=\\"off\\" class=\\"vs__search\\"></div>
      <div class=\\"vs__actions\\"><button type=\\"button\\" title=\\"Clear Selected\\" aria-label=\\"Clear Selected\\" class=\\"vs__clear\\" style=\\"display: none;\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"10\\" height=\\"10\\">
            <path d=\\"M6.895455 5l2.842897-2.842898c.348864-.348863.348864-.914488 0-1.263636L9.106534.261648c-.348864-.348864-.914489-.348864-1.263636 0L5 3.104545 2.157102.261648c-.348863-.348864-.914488-.348864-1.263636 0L.261648.893466c-.348864.348864-.348864.914489 0 1.263636L3.104545 5 .261648 7.842898c-.348864.348863-.348864.914488 0 1.263636l.631818.631818c.348864.348864.914773.348864 1.263636 0L5 6.895455l2.842898 2.842897c.348863.348864.914772.348864 1.263636 0l.631818-.631818c.348864-.348864.348864-.914489 0-1.263636L6.895455 5z\\"></path>
          </svg></button> <span></span>
        <div class=\\"vs__spinner\\" style=\\"display: none;\\">Loading...</div>
      </div>
    </div>
    <ul id=\\"vs1__listbox\\" role=\\"listbox\\" style=\\"display: none; visibility: hidden;\\"></ul>
  </div>
</div>"
`;
